clippy_lints/src/mem_replace.rs

```diff
@@ -200,7 +200,22 @@ fn check_replace_with_default(cx: &LateContext<'_>, src: &Expr<'_>, dest: &Expr<
             if !in_external_macro(cx.tcx.sess, expr_span);
             if let ExprKind::Path(ref repl_func_qpath) = repl_func.kind;
             if let Some(repl_def_id) = cx.qpath_res(repl_func_qpath, repl_func.hir_id).opt_def_id();
-            if match_def_path(cx, repl_def_id, &paths::DEFAULT_TRAIT_METHOD);
+
+            let defaults = &[
+                paths::DEFAULT_TRAIT_METHOD.as_ref(),
+                paths::STRING_NEW.as_ref(),
+                paths::VEC_NEW.as_ref(),
+                paths::VEC_DEQUE_NEW.as_ref(),
+                paths::LINKED_LIST_NEW.as_ref(),
+                paths::HASHMAP_NEW.as_ref(),
+                paths::BTREEMAP_NEW.as_ref(),
+                paths::HASHSET_NEW.as_ref(),
+                paths::BTREESET_NEW.as_ref(),
+                paths::BINARY_HEAP_NEW.as_ref(),
```

Comment:
```suggestion
                (sym::BinaryHeap, sym::new)),
```

All of these types will have diagnostic items after the next sync with rustc so we should probably use those instead of adding more paths. You can use the util function `is_diagnostic_assoc_item` instead of `match_def_path`.

Reply:
I wasn't aware of it, thanks! Makes sense to me to wait then.
I got a question related to this:
- For `Default` trait, won't we need `default_trait` symbol specified (similar to e.g. [debug_trait](https://github.com/rust-lang/rust/blob/master/library/core/src/fmt/mod.rs#L549))? It is not defined for the [Default trait](https://github.com/rust-lang/rust/blob/master/library/core/src/default.rs#L84).  Or maybe `Default` symbol is sufficient?

Reply:
Oh that one needs to be added. Or use the current method for now. The symbol needed just depends on what the diagnostic item is named in `#[rustc_diagnostic_item = "name"]`. Personally I like naming it the same as the item like `Default`.

Reply:
I see that the name we need to use comes from `#[rustc_diagnostic_item = "name"]`, but by that:
> Or maybe `Default` symbol is sufficient?

I meant already defined `Default` symbol in [symbol.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/symbol.rs#L143). I don't quite understand what it is then (because `#[rustc_diagnostic_item="..."` is not defined for the `Default` trait)

Reply:
A Symbol is just an interned string. So if we add `#[rustc_diagnostic_item="Default"]` and `Default` is already in symbol.rs then we don't need to add to symbol.rs. We also could use `sym!(Default)` which expands to `Symbol::intern("Default")`. But having `sym::Default` is better because it is const. I hope that answers your question?

Reply:
Now it is more clear, thanks!
Created [this PR in rust repo](https://github.com/rust-lang/rust/pull/82773) to add diagnostic item to `Default` trait.